### ISSUE TITLE
Redundant short isoforms not properly collapsed?

### ISSUE THREAD

**Issue by SziKayLeung** [Magdoll/cDNA_Cupcake#129]
Hello Liz, 

Hope you are well. 

I'm using the default parameters for collapse_isoforms_by_sam.py to collapse my ERCC standards, and find that several ERCCs are annotated to several isoforms. 

The fasta sequence of the isoforms (see attached, collapsed.filtered.rep.fq) classified as unique but mapped to the same ERCC have the same exact sequence - only difference is in length (some of the isoforms are shorter and truncated at the 3' end). My assumption is that they are the same isoform (especially since it's ERCCs). 
I've played around with --max_3_diff, but made no difference on collapse. 

Your thoughts and help on this would be greatly appreciated - thank you!

Szi Kay
[7isoforms_ERCC130.txt](https://github.com/Magdoll/cDNA_Cupcake/files/5275882/7isoforms_ERCC130.txt)

**Comment by Magdoll** [comment]
Hi @SziKayLeung ,
Can you please provide me the full collapse command you are using?  And it's the latest Cupcake versions right?
THanks
-Liz

**Comment by SziKayLeung** [comment]
Thanks Liz for a speedy response.
The collapse command: python collapse_isoforms_by_sam.py --input K18.polished.hq.fastq --fq -s K18.polished.hq.fastq.sorted.sam -o K18; 
And checked that it's the latest version (v12)

**Comment by SziKayLeung** [comment]
Hello Liz, 

I went back to the fasta sequence generated from CCS and LIMA to check whether these ‘isoforms’ from the previous attached document have poly-A tail, and indeed they do…for example, two of ‘isoforms’ from ERCC-00130 had FL reads with the same sequence (and that of the original ERCC sequence), but with different lengths and both ending with polyA. 

I’m not sure why there are random polyA sequences (random priming oligo-DT??), but it appears to be an issue already upstream of cupcake scripts – sorry. I guess this is not something that can be automatically detected from playing around with the current parameters for collapse_isoforms_by_sam.py 

On a side note, I wonder if this is something that you have seen before? 
Thank you
[PolyA_ERCC130.txt](https://github.com/Magdoll/cDNA_Cupcake/files/5277188/PolyA_ERCC130.txt)

**Comment by Magdoll** [comment]
Hi @SziKayLeung ,
I was able to actually collapse all of the ERCC-130 isoforms into a single representation...is that not the intended outcome?

I used minimap2 to map, and notice there were soft-clippings (untrimmed sequences in the beginning?) so I adjusted the alignment query coverage parameter (`-c`) down to 0.95 and then was able to collapse everything. I ran it with two parameter sets that you can find in the [dropbox tar here](https://www.dropbox.com/s/k51lcopirfr3nqq/Oct2020_Cupcake_issue129_SziKay_ERCC.tar.gz?dl=0)

**Comment by SziKayLeung** [comment]
Hi @Magdoll , 
Thank you for rerunning the code, and yes! The intended outcome is to collapse all of the ERCCs into one isoform. 

However, I noticed that your input fasta was actually the output fasta from running the Cupcake collapse rather than from the polished fasta file from Iso-Seq Cluster. 
Attempting the same line of thought, I was only able to re-collapse all the ERCCs by running Cupcake twice i.e., 
1) Iso-Seq3 pipeline 
2) Minimap2, input = polished.fa from Iso-Seq Cluster  
3) Cupcake, input = polished.fa from Iso-Seq Cluster, output = collapsed.rep.fa
4) Minimap2, input = collapsed.rep.fa 
5) Cupcake, input = collapsed.rep.fa
(I've attached the initial output [hq, polished fasta file](https://www.dropbox.com/s/rb14egepal43p35/Oct2020_Cupcake_issue129.zip?dl=0) from Iso-Seq Cluster for reference) 


I'm happy to close the query as this is what I was hoping, but do let me know your thoughts on rerunning cupcake twice. Thank you!
